qutip/control/pulsegen.py

```diff
@@ -234,50 +236,120 @@ def init_pulse(self):
 
         self._pulse_initialised = True
 
-        if not self.lbound is None:
-            if np.isinf(self.lbound):
-                self.lbound = None
-        if not self.ubound is None:
-            if np.isinf(self.ubound):
-                self.ubound = None
-
-        if not self.ubound is None and not self.lbound is None:
-            if self.ubound < self.lbound:
+        if (
+            isinstance(self.lbound, np.ndarray) or
+            isinstance(self.ubound, np.ndarray)
+        ):
+            if not (
+                isinstance(self.lbound, np.ndarray) and
+                isinstance(self.ubound, np.ndarray)
+            ):
+                raise ValueError(
+                    "If either lbound or ubound is an array, both must be."
+                )
+            if self.lbound.shape != self.ubound.shape:
+                raise ValueError(
+                    "If lbound and ubound are arrays, they should have the"
+                    " same shape"
+                )
+            if None in self.lbound or None in self.ubound:
+                raise ValueError(
+                    "If lbound and ubound are arrays, use np.inf or -np.inf,"
+                    " not None, to specify the lack of a bound."
+                )
+            if (self.ubound < self.lbound).any():
                 raise ValueError("ubound cannot be less the lbound")
-
-    def _apply_bounds_and_offset(self, pulse):
+        else:
+            if isinstance(self.lbound, float):
+                if np.isinf(self.lbound):
+                    self.lbound = None
+            if isinstance(self.ubound, float):
+                if np.isinf(self.ubound):
+                    self.ubound = None
+            if self.ubound is not None and self.lbound is not None:
+                if self.ubound < self.lbound:
+                    raise ValueError("ubound cannot be less the lbound")
+
+    def _apply_bounds_and_offset(self, pulse, ctrl_index=None):
         """
         Ensure that the randomly generated pulse fits within the bounds
         (after applying the offset)
         Assumes that pulses passed are centered around zero (on average)
+        ctrl_index is the index of the control for which the pulse is computed
         """
         if self.lbound is None and self.ubound is None:
             return pulse + self.offset
 
-        max_amp = max(pulse)
-        min_amp = min(pulse)
-        if ((self.ubound is None or max_amp + self.offset <= self.ubound) and
-            (self.lbound is None or min_amp + self.offset >= self.lbound)):
-            return pulse + self.offset
+        if (
+            # init_pulse should ensure that these are either both ndarrays
+            # or both not ndarrays
+            isinstance(self.lbound, np.ndarray) or
+            isinstance(self.ubound, np.ndarray)
+        ):
+            if ctrl_index not in range(self.lbound.shape[1]):
```

Comment:
I think the check is okay (`ctrl_index` could be something weird like `None`) but maybe the error message isn't good enough (since it assumes `ctrl_index` is fine).